Sources/Credentials/CredentialsTokenTTL.swift

```diff
@@ -0,0 +1,117 @@
+    /// - Parameter options: The dictionary of plugin specific options.
+    func generateNewProfile(token: String, options: [String:Any], completion: @escaping (CredentialsTokenTTLResult) -> Void)
+}
+
+/// Enum with cases corresponding to the onSuccess and onFailure closures in the authentication method.
+public enum CredentialsTokenTTLResult {
+    case success(UserProfile)
+    case failure(HTTPStatusCode?, [String:String]?)
+    
+    /// Helper method to convert an Error to a failure enum
+    public static func error(_ error: Swift.Error) -> CredentialsTokenTTLResult {
+        return .failure(nil, ["failure": "\(error)"])
+    }
+}
+
+extension CredentialsTokenTTL {
+    func getProfileFromCache(token: String) -> UserProfile? {
+        #if os(Linux)
+            let key = NSString(string: token)
+        #else
+            let key = token as NSString
+        #endif
+        
+        if let cached = usersCache?.object(forKey: key) {
+            if let ttl = tokenTimeToLive {
+                if Date() < cached.createdAt.addingTimeInterval(ttl) {
+                    return cached.userProfile
+                }
+                // If current time is later than time to live, continue to standard token authentication.
+                // Don't need to evict token, since it will replaced if the token is successfully authenticated.
+            } else {
+                // No time to live set, use token until it is evicted from the cache
+                return cached.userProfile
+            }
+        }
+        
+        return nil
+    }
+    
+    func saveProfileToCache(token: String, profile: UserProfile) {
+        let newCacheElement = BaseCacheElement(profile: profile)
+        #if os(Linux)
+            let key = NSString(string: token)
+        #else
+            let key = token as NSString
+        #endif
+        
+        self.usersCache!.setObject(newCacheElement, forKey: key)
+    }
+
+    /// Calls the completion handler with the profile (from cache or generated with the protocol generateNewProfile method), or failure result.
+    ///
+    /// - Parameter token: The Oauth2 token, used as a key in the cache.
+    /// - Parameter options: The dictionary of plugin specific options.
+    /// - Parameter onSuccess: From the authentication method.
+    /// - Parameter onFailure: From the authentication method.
+    ///
+    public func getProfileAndCacheIfNeeded(
+        token: String,
+        options: [String:Any],
+        onSuccess: @escaping (UserProfile) -> Void,
+        onFailure: @escaping (HTTPStatusCode?, [String:String]?) -> Void) {
+        
+        if let profile = getProfileFromCache(token: token) {
+            onSuccess(profile)
+            return
+        }
+        
+        // Either the token/profile expired or there was none in the cache. Make one.
+        
+        generateNewProfile(token: token, options: options) {[weak self] generatedResult in
+            guard let self = self else {
```

Comment:
Hmmm. I have changed this to see if it helps. However, I have been using this pattern for a while. For example, it's in one of my production apps on the Apple App Store (see https://github.com/crspybits/SharedImages/blob/master/Code/Media/Media%20Views/URLMediaView.swift).

Reply:
We'll, that fixed the CI issue. Interesting.

Reply:
I suspect it's something that changed in later versions of Swift.  Kitura packages still support all the way back to Swift 4.0.3, so the language/compiler improvements since then are not available.